src/core/thread_pool.h

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include <condition_variable>
+#include <deque>
+#include <future>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+#include "src/core/cuda_utils.h"
+#include "src/core/status.h"
+#include "src/core/sync_queue.h"
+
+namespace nvidia { namespace inferenceserver {
+
+class WorkerThreadPool {
+ public:
+  WorkerThreadPool(int thread_count) : thread_count_(thread_count)
+  {
+    worker_threads_.reserve(thread_count);
+    futures_.reserve(thread_count);
+  }
+
+  ~WorkerThreadPool();
+
+  // Wait for all pending worker threads to finish.
+  Status AwaitCompletion();
+
+  // Get Id of next available worker thread. If all workers are occupied then
+  // wait for them to finish and start from the beginning.
+  Status GetNextAvailableId(int* worker_id);
+
+  // Add CopyBuffer task to queue
+  Status AddTask(
+      const std::string& msg, const TRITONSERVER_MemoryType src_memory_type,
+      const int64_t src_memory_type_id,
+      const TRITONSERVER_MemoryType dst_memory_type,
+      const int64_t dst_memory_type_id, const size_t byte_size, const void* src,
+      void* dst, cudaStream_t cuda_stream, bool* cuda_used);
+
+  // Run CopyBuffer on the worker threads. Must add to queue using AddTask
+  // before calling ProcessQueue.
+  Status ProcessQueue();
```

Comment:
Can you illustrate the intended workflow of using this thread pool class? I don't think the public functions except `AddTask()` are necessary.
Also who is responsible for calling `ProcessQueue()`? It simply tells worker threads to consume the tasks in the queue, why don't you just do the same when invoking `AddTask()`

Reply:
I think you can just re-purpose this function to be waiting for all tasks are completed? And then once this function returns, the state within the class object is reset for next set of tasks.

Reply:
AddTask will add the CopyBuffer task to the queue. 
To execute the tasks we must call ProcessQueue. The ProcessQueue returns the status (success or failure + failure msg) once all the tasks are completed.
Are you suggested we start processing the task as soon as it is added to the queue instead of waiting for all the tasks to be added to the queue?

Reply:
If the number of tasks > number of worker threads, we may have to wait at AddTask for a thread to be available so that it can trigger the next CopyBuffer op.

Reply:
> Are you suggested we start processing the task as soon as it is added to the queue instead of waiting for all the tasks to be added to the queue?

Yes, I don't see a reason for worker thread to wait until it is told to consume the queue, which introduce unnecessary wait, it can just actively wait for task inside the queue.

> If the number of tasks > number of worker threads, we may have to wait at AddTask for a thread to be available so that it can trigger the next CopyBuffer op.

Please take a look at dynamic batcher or plan backend for how sync_queue is being used. The producer doesn't need to worry about the consumer, it just places tasks into the queue.

Reply:
Fixed.